File: @dayzcc/install/migrate.pl
#!/usr/bin/perl -w
# Database schema migration utility 
# by ayan4m1 and Crosire

use Getopt::Long;
use File::Basename;
use DBIx::Migration::Directories;
use DBIx::Transaction;
use DBI;

our %args;
GetOptions(
	\%args,
	'hostname|host|h=s',
	'username|user|u=s',
	'password|pass|dbpass|p=s',
	'database|name|dbname|d=s',
	'port|dbport=s',
	'help'
);

my %db = (
	'host' => $args{'hostname'} ? $args{'hostname'} : '127.0.0.1',
	'user' => $args{'username'} ? $args{'username'} : 'dayz',
	'pass' => $args{'password'} ? $args{'password'} : '',
	'name' => $args{'database'} ? $args{'database'} : 'dayz_chernarus',
	'port' => $args{'port'} ? $args{'port'} : '3306'
);

if ($args{'help'}) {
	print "Usage: migrate.pl [--host <hostname>] [--user <username>] [--pass <password>] [--name <dbname>] [--port <port>] [--schema <schema>] [--version <version>]\n";
	exit;
}

print "Starting up ...\n";
die "> Error: Schema version must be specified for a non-standard schema!\n" if ($args{'schema'} && !defined $args{'version'});
print "\n";

my $schema1 = "Bliss";
my $schema2 = "Controlcenter";
my $version1 = "0.36";
my $version2 = "0.08";

print "Trying to connect to MySQL server on $db{'host'}, database $db{'name'} as $db{'user'}\n";
my $dbh = DBIx::Transaction->connect("dbi:mysql:$db{'name'}:$db{'host'}:$db{'port'}", $db{'user'}, $db{'pass'}) or die "> Error: MySQL Error: ".DBI->errstr."\n";
print "> Sucessfully connected!\n\n";

my $m1 = DBIx::Migration::Directories->new(base => dirname(__FILE__).'/schema', schema => $schema1, desired_version => $version1, dbh => $dbh);
my $m2 = DBIx::Migration::Directories->new(base => dirname(__FILE__).'/schema', schema => $schema2, desired_version => $version2, dbh => $dbh);
if ($m1->get_current_version()) { printf("Current ".$schema1." schema version is %.2f\n", $m1->get_current_version()); } else { print "No existing ".$schema1." schema was found!\n"; }
if ($m2->get_current_version()) { printf("Current ".$schema2." schema version is %.2f\n", $m2->get_current_version()); } else { print "No existing ".$schema2." schema was found!\n"; }
print "\n";

print "Running database migration ...\n\n";
$m1->migrate or die "> Error: Migration failed!\n";
$m2->migrate or die "> Error: Migration failed!\n";
print "\n";
print "> Completed ".$schema1." update to version $version1\n";
print "> Completed ".$schema2." update to version $version2\n";
print "\n";

print "Running optional package migration ...\n\n";
$m1 = DBIx::Migration::Directories->new(base => dirname(__FILE__).'/schema', schema => "BlissBuildings", desired_version => "0.01", dbh => $dbh);
$m1->migrate;
$m1 = DBIx::Migration::Directories->new(base => dirname(__FILE__).'/schema', schema => "BlissInvCust", desired_version => "0.02", dbh => $dbh);
$m1->migrate;
print "\n> Completed optional package update";

File: @dayzcc/install/migrate.php
<?php

/**
 * Tiny database migration script by Crosire
 */

chdir(dirname(__FILE__));

$args = getopt("", array("host:", "port:", "username:", "password:", "database:", "schema:"));
$path_schema = "schema";
$path_mysql = "..\\mysql\\bin\\mysql.exe";

$dbhost = (isset($args['host']) ? $args['host'] : "127.0.0.1");
$dbport = (isset($args['port']) ? $args['port'] : "3306");
$dbuser = (isset($args['username']) ? $args['username'] : "dayz");
$dbpass = (isset($args['password']) ? $args['password'] : "");
$dbname = (isset($args['database']) ? $args['database'] : "dayz_chernarus");
$schema = (isset($args['schema']) ? $args['schema'] : "Bliss");
$version = "0.00";

// Connect to the MySQL server
echo "Connecting to MySQL server on ".$dbhost." as ".$dbuser." to database ".$dbname."\n";
$link = mysql_connect($dbhost, $dbuser, $dbpass);
if (!$link) {
	echo "> Could not connect: ".(mysql_error())."\n";
	exit(1);
}
mysql_select_db($dbname, $link);
echo "> Sucessfully connected!\n\n";

// Get the last version
$res = mysql_query("SELECT `version` FROM `migration_schema_version` WHERE `name` = '".$schema."'", $link);
if (!$res) {
	echo "No existing ".$schema." schema was found!\n";
} else {
	$version = rtrim(mysql_fetch_assoc($res)['version'], "0");
	echo "Current ".$schema." version is: $version\n";
}

function get_folders() {
	global $path_schema, $schema;
	$folders = array();
	foreach(scandir($path_schema."\\".$schema."\\mysql") as $dir) {
		if (is_dir($path_schema."\\".$schema."\\mysql\\".$dir)) {
			if (preg_match("/([.0-9]{4})-?([.0-9]{4})?/", $dir, $matches)) {
				if (isset($matches[2])) { $folders[] = array($matches[2], $dir); } else { $folders[] = array($matches[1], $dir); };
			}
		}
	}
	return $folders;
}
function get_files($dir) {
	global $path_schema, $schema;
	$files = array();
	$handle = opendir($path_schema."\\".$schema."\\mysql\\".$dir);
	while (false !== ($file = readdir($handle))) {
		if (preg_match("/([0-9]+)_([_a-zA-Z0-9]*).sql/", $file)) {
			$files[] = array($file, $path_schema."\\".$schema."\\mysql\\".$dir."\\".$file);
		}
	}
	closedir($handle);
	return $files;
}

// Run migration proces
echo "\nRunning database migration ...\n";
$res = mysql_query("CREATE DATABASE IF NOT EXISTS `".$dbname."`", $link);
if (!$res) {
	echo "> Database Creation failed: ".(mysql_error())."\n";
	exit(2);
}
mysql_select_db($dbname, $link);
mysql_query("GRANT ALL PRIVILEGES ON ".$dbname.".* TO 'dayz'@'localhost' IDENTIFIED BY 'dayz'", $link);
mysql_query("GRANT ALL PRIVILEGES ON mysql.* TO 'dayz'@'localhost'", $link);

// Execute sql files in schema folder
foreach (get_folders() as $folder) {
	if ($folder[0] > $version) {
		foreach (get_files($folder[1]) as $file) {
			echo "> ".$folder[1]."\\".$file[0]."\n";
			$output = shell_exec($path_mysql." --host={$dbhost} --port={$dbport} --user={$dbuser} --password={$dbpass} {$dbname} < {$file[1]}");
		}
		$oldversion = $version;
		$version = $folder[0];
		mysql_query("INSERT INTO `migration_schema_log` (`schema_name`, `event_time`, `old_version`, `new_version`) VALUES ('".$schema."', NOW(), '".$oldversion."', '".$version."')", $link);
	}
}

// Output the new version number and exit
mysql_query("INSERT INTO `migration_schema_version` (`name`, `version`) VALUES ('".$schema."', '".$version."') ON DUPLICATE KEY UPDATE `version` = '".$version."'", $link);
mysql_close($link);

echo "> Completed ".$schema." update to version ".$version."\n";
exit;

?>